### PR TITLE
add Collector.Default

### PR DESCRIPTION
**Pull request by c-cube**
- add `add_data_to_span` and `add_data_to_explicit_span`
- update test
- add Collector.Default

### CODE REVIEW

File: src/core/collector.ml [hunk 1]
```diff
@@ -53,6 +53,10 @@ module type S = sig
   (** Exit an explicit span.
       @since 0.3 *)
 
+  val add_data_to_span : span -> (string * user_data) list -> unit
+  (** @since Adds data to the given span.
+      NEXT_RELEASE *)
+
```
Comment (ELLIOTTCABLE) [inline src/core/collector.ml:59]: See [my review](https://github.com/c-cube/ocaml-trace/pull/11#pullrequestreview-1601545377) of #11

File: src/core/collector.ml [hunk 2]
```diff
@@ -71,3 +75,27 @@ module type S = sig
   val shutdown : unit -> unit
   (** Shutdown collector, possibly waiting for it to finish sending data. *)
 end
+
+(** Default implementation that does nothing. Collectors that wish to
+    be more resilient in the face of futures additions to {!S}
+    can start with [include Trace.Collector.Default]. *)
+module Default : S = struct
+  let _fail what =
+    failwith (Printf.sprintf "not implemented: Trace.Collector.%s" what)
```
Comment (ELLIOTTCABLE) [inline src/core/collector.ml:84]: Perhaps it's a good idea to mirror my implementation, and require collectors expose a string-ish `name` that can be printed in debugging contexts like this one?

https://github.com/ELLIOTTCABLE/ocaml-ambient-context/blob/30073325fe8d168ce44ac06b8b338d5b4875b02c/lib/types.ml

Comment (ELLIOTTCABLE) [reply src/core/collector.ml:84]: In fact, perhaps this should even be abstracted — in additional to `message`, add `internal_message`, which `ocaml-trace` itself will call in erroneous situations? This allows collectors to self-collect.

**Review by ELLIOTTCABLE — CHANGES_REQUESTED** [review]

**Comment by c-cube** [comment]
Guess it's not worth it. Collectors will just have to depend on a more specific version of ocaml-trace.

**Pull request closed without merging**